File: legacy/src/main/rs/emulate/legacy/config/DefaultConfigDefinition.java
package rs.emulate.legacy.config;

import com.google.common.collect.ImmutableMap;

import java.util.Map;

/**
 * A base class for default definitions. All default definitions <strong>must</strong> be <strong>immutable</strong>.
 *
 * @param <T> The type of definition this default is for.
 * @author Major
 */
public abstract class DefaultConfigDefinition<T extends MutableConfigDefinition> {

	/**
	 * The map of opcodes to DefinitionProperty objects.
	 */
	private final Map<Integer, SerializableProperty<?>> properties;

	/**
	 * Creates the DefaultDefinition.
	 */
	protected DefaultConfigDefinition() {
		this.properties = ImmutableMap.copyOf(init());
	}

	/**
	 * Gets a {@link ConfigPropertyMap}
	 *
	 * @return The PropertyMap.
	 */
	public final ConfigPropertyMap toPropertyMap() {
		return new ConfigPropertyMap(properties);
	}

	/**
	 * Initialises the DefaultDefinition.
	 *
	 * @return The {@link Map} of opcodes to {@link SerializableProperty} objects.
	 */
	protected abstract Map<Integer, SerializableProperty<?>> init();

}

File: legacy/src/main/kotlin/rs/emulate/legacy/config/DefaultConfigDefinition.kt
package rs.emulate.legacy.config

import com.google.common.collect.ImmutableMap

/**
 * A base class for default definitions. All default definitions **must** be immutable.
 *
 * @param T The type of definition this default is for.
 */
abstract class DefaultConfigDefinition<T : MutableConfigDefinition> protected constructor() {

    /**
     * The map of opcodes to DefinitionProperty objects.
     */
    private val properties: Map<Int, SerializableProperty<*>> = ImmutableMap.copyOf(init()) // TODO by lazy

    /**
     * Gets a [ConfigPropertyMap]
     */
    fun toPropertyMap(): ConfigPropertyMap {
        return ConfigPropertyMap(properties)
    }

    /**
     * Initialises the DefaultDefinition.
     *
     * @return The [Map] of opcodes to [SerializableProperty] objects.
     */
    protected abstract fun init(): Map<Int, SerializableProperty<*>>

}
